File: plugins/typescript/examples/json-demo.ts
#!/usr/bin/env node
// Example using the optional structured-JSON protocol instead of the text
// format. A plugin opts in by printing a `{"vee":1,…}` object.
//
// `JSONMenu` mirrors `Menu` method for method, so the only thing that changes
// between the two wire formats is which builder you construct. Its `build()`
// output is committed to plugins/fixtures/json-demo.txt and checked for drift.
//
// Using this outside the repository: run `vee sdk ts` to write vee.ts beside
// your copy, then change the import below to "./vee.ts".
import { fileURLToPath } from "node:url";
import { JSONMenu } from "../vee.ts";

export function build(): string {
  const menu = new JSONMenu();
  menu.title("JSON ✓", { color: "green", sfimage: "curlybraces" });

  const d = menu.dropdown;
  d.item("Structured item", { href: "https://example.com" });
  d.separator();
  d.submenu("Submenu").item("Child", { color: "blue" });
  // Characters the three languages' JSON encoders disagree about by default:
  // Python escapes non-ASCII, Go escapes <, > and & for HTML embedding. Pinned
  // here so any of them regressing fails the drift guard.
  d.item("R&D <beta> ✓", { tooltip: "a & b" });
  return menu.toString();
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  process.stdout.write(build() + "\n");
}

File: plugins/python/examples/json-demo.py
#!/usr/bin/env python3
# The structured-JSON protocol, built with JSONMenu. Byte-identical to the
# TypeScript and Go examples; committed to plugins/fixtures/json-demo.txt.
#
# Using this outside the repository: run `vee sdk py` to write vee.py beside
# your copy. No edit needed -- Python already searches the script's own
# directory.
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from vee import JSONMenu  # noqa: E402


def build() -> str:
    menu = JSONMenu()
    menu.title("JSON ✓", color="green", sfimage="curlybraces")

    d = menu.dropdown
    # Surface targeting, in its JSON spelling: the same two axes the text
    # protocol writes as `visibleon=`/`searchable=` (see the surfaces example).
    d.item("Structured item", href="https://example.com",
           visible_on=["menu", "window"], searchable=False)
    d.separator()
    d.submenu("Submenu").item("Child", color="blue")
    # See the TypeScript example: characters the three JSON encoders disagree
    # about by default.
    d.item("R&D <beta> ✓", tooltip="a & b")
    return menu.to_string()


if __name__ == "__main__":
    print(build())
